canon.go

```diff
@@ -91,3 +91,38 @@ func canonValue(chain *chain, in interface{}) (interface{}, bool) {
 
 	return out, true
 }
+
+func canonDecode(chain *chain, value interface{}, target interface{}) {
+	if chain.failed() {
+		return
+	}
+
+	if target == nil {
+		chain.fail(AssertionFailure{
+			Type:   AssertNil,
```

Comment:
AssertNil -> AssertUsage (we use this kind of assertion for cases when the lib is used incorrectly, like here)

Reply:
Oh ok, when do we use AssertNil then?

Reply:
```
FAIL: TestArray_Decode/Target_is_nil (0.00s)
panic: AssertionFailure of type AssertUsage can't have Actual field [recovered]
        panic: AssertionFailure of type AssertUsage can't have Actual field
```

I'm guessing we can't use AssetUsage without an actual value, which in our case is nil.

Reply:
Usually we use them like this:

* if user asks that value should be nil: AssertNil
* if user asks that value should not be nil: AssertNotNil
* if user incorrectly uses the library: AssertUsage

In this case, the *user* does not want to check `target` for null, the user just requests some operation that involves `target`. And it's *our* requirement that the `target` is not null, and if it is, we want to tell the user that the library is being used incorrectly and we can't fulfill the user request.

Reply:
This is of course just a convention and AssertNotNil would work too. But use AssertUsage in similar places when the user provides some incorrect arguments, and it's handy that all such places can be easily identified by AssertUsage.

Reply:
Ahh, I see!!.